### PR TITLE
DELETE search route

### PR DESCRIPTION
**Pull request by ereteog**
> Close #https://github.com/threatgrid/iroh/issues/4385

This PR introduces a `DELETE /ctia/{entity}/search` route that enables users to delete entities that match given filters.
The search filters are the same than those of the `search` / `count`and metrics routes.
In order to prevent (as far as possible) deletion errors, the route triggers a dry run by default that only returns the number of matched entities and empty queries (that would match all entities) are forbidden. One can still delete all it's visible entities by using the `query` filter with `*`.
I also reduced some boilerplate in access control tests and improved es store crud tests.

Reviewers: you should hide whitespace changes.

<a name="qa">[§](#qa)</a> QA
============================

let 3 users
user1 in group1 with search and delete capabilities
user2 is in group1 with search and delete capabilities
user3 is in group1 with only search capability
user4 is in group2 with search and delete capabilities

1) user 1 post a bundle with 3 sightings green, 2 sighting amber, 1 sighting red

2) user4 makes a `DELETE /ctia/sighting/search` query that matches all created sightings with `REALLY_DELETE_ALL_THOSE_ENTITIES` parameter set `true`.
==> Check that the response is 0 and that no entities were deleted.

3) user2 makes a `DELETE /ctia/sighting/search` query that matches all created sightings with `REALLY_DELETE_ALL_THOSE_ENTITIES` parameter set `false`.
==> Check that the response is 5 and that no entities were deleted.

4) user2 makes a `DELETE /ctia/sighting/search` query that matches all created sightings without `REALLY_DELETE_ALL_THOSE_ENTITIES` parameter.
==> Check that the response is 5 and that no entities were deleted.

5) user2 makes a `DELETE /ctia/sighting/search` without any filters and `REALLY_DELETE_ALL_THOSE_ENTITIES` parameter et to true.
==> Check that the response has status 403 and that no entities were deleted.

6) user3 makes a `DELETE /ctia/sighting/search` query that matches all created sightings with `REALLY_DELETE_ALL_THOSE_ENTITIES` parameter set to `true`.
==> Check that the response has status 401 and that no entities were deleted.

7) user2 makes a `DELETE /ctia/sighting/search` query that matches all created sightings with `REALLY_DELETE_ALL_THOSE_ENTITIES` parameter set to `true`.
==> Check that the response is 5 and that only the 3 green and 2 amber entities were deleted.

8) user1 makes a `DELETE /ctia/sighting/search` query that matches all created sightings with `REALLY_DELETE_ALL_THOSE_ENTITIES` parameter set to `true`.
==> Check that the response is 1 and that only the remaining 2 red entities were deleted.

<a name="release-notes">[§](#release-notes)</a> Release Notes
=============================================================

```
intern: introduced DELETE search route.
```

### CODE REVIEW

File: src/ctia/task/purge_es_stores.clj [hunk 1]
```diff
@@ -1,40 +0,0 @@
-(ns ctia.task.purge-es-stores
```
Comment (ereteog) [inline src/ctia/task/purge_es_stores.clj:1]: this is not maintained nor used anymore.
One can simply launch 
* `DELETE http://localhost:9200/ctia_*` to delete CTIA indices
* `DELETE http://localhost:9200/_template/ctia_*` to delete CTIA indices

File: src/ctia/http/routes/crud.clj [hunk 2]
```diff
@@ -310,34 +311,74 @@
...
+                                              " the inevitable disaster that will occur after"
+                                              " you perform that operation."
+                                              " DO NOT FORGET TO SET THAT TO FALSE AFTER EACH DELETION"
+                                              " IF YOU INTEND TO USE THAT ROUTE MULTIPLE TIMES."))})]
+             (let [query (->> (dissoc params :wait_for :REALLY_DELETE_ALL_THOSE_ENTITIES)
+                              (search-query date-field))]
+               (if (empty? query)
+                 (forbidden {:error "you must provide at least one of from, to, query or any field filter."})
+                 (ok
+                  (if (:REALLY_DELETE_ALL_THOSE_ENTITIES params)
```
Comment (frenchy64) [inline src/ctia/http/routes/crud.clj:370]: `THOSE` => `THESE`

File: test/ctia/test_helpers/es.clj [hunk 3]
```diff
@@ -256,10 +256,14 @@
...
-           (testing (format "%s (ES version: %s)." msg#  ~'version)
+           (testing (format "%s (ES version: %s).\n" msg#  ~'version)
              (when clean-fn#
                (clean-fn# ~'conn))
-             ~@body)
+             (h/with-properties*
+               ["ctia.store.es.default.port" ~'es-port
+                "ctia.store.es.default.version" ~'version]
+               (fn []
+                 ~@body)))
```
Comment (frenchy64) [inline test/ctia/test_helpers/es.clj:266]: I'd recommend using `(h/with-properties [...] ~@body)` (the macro) or wrapping the body in a `do` `(h/with-properties* [...] (fn [] (do ~@body))` to distinguish from the :pre/post syntax.

**Review by frenchy64 — APPROVED** [review]
LGTM!

File: src/ctia/entity/event/store.clj [hunk 4]
```diff
@@ -14,6 +14,7 @@
     (crud/handle-create state new-events))
   (list-events [this filter-map ident params]
     (crud/handle-list state filter-map ident params))
+  (close [_] (close-cm! state))
```
Comment (agzam) [inline src/ctia/entity/event/store.clj:17]: Perhaps, at some point we could discuss renaming `ctia.stores.es.store/close-cm!`. Arguably, `cm` maybe not very obvious abbreviation,  without jumping to where the fn is implemented, it's not clear. But, anyone who reads the code has to do that only once, so, really not that big of a deal.

**Review by agzam — APPROVED** [review]
Looks good.

File: src/ctia/entity/event/store.clj (hunk 4 above)
Comment (ereteog) [reply src/ctia/entity/event/store.clj:17]: I agree. I always prefer easier code to read. I replaced it by `close-connections` (indeed it closes a connection manager).

**Comment by NeelSaini** [comment]
verified in rel v1.63

**Pull request merged**